Sources/SwiftDiagnostics/DiagnosticsFormatter.swift

```diff
@@ -103,12 +103,76 @@ public struct DiagnosticsFormatter {
         }
 
         for diag in diags.dropLast(1) {
-          annotatedSource.append("\(preMessage)├─ \(diag.message)\n")
+          if colorize {
+            annotatedSource.append("\(preMessage)├─ \(self.colorize(diag.diagMessage))\n")
+          } else {
+            annotatedSource.append("\(preMessage)├─ \(diag.message)\n")
+          }
         }
-        annotatedSource.append("\(preMessage)╰─ \(diags.last!.message)\n")
 
+        if colorize {
+          annotatedSource.append("\(preMessage)╰─ \(self.colorize(diags.last!.diagMessage))\n")
+        } else {
+          annotatedSource.append("\(preMessage)╰─ \(diags.last!.message)\n")
+        }
       }
     }
     return annotatedSource
   }
+
+  /// Annotates the given ``DiagnosticMessage`` with an appropriate ANSI color code and returns the result as a printable string.
+  static func colorize(_ message: DiagnosticMessage) -> String {
+    switch message.severity {
+    case .error:
+      let annotation = ANSIAnnotation(color: .red, trait: .bold)
+      return annotation.applied(to: message.message)
+    case .warning:
+      let annotation = ANSIAnnotation(color: .yellow)
+      return annotation.applied(to: message.message)
+    case .note:
+      return message.message
+    }
+  }
+
+  enum ANSIColor: UInt8 {
+    case reset = 0
+    case black = 30
+    case red = 31
+    case green = 32
+    case yellow = 33
+    case blue = 34
+    case magenta = 35
+    case cyan = 36
+    case white = 37
+  }
+
+  enum ANSITrait: UInt8 {
+    case bold = 1
+    case underline = 4
+  }
+
+  struct ANSIAnnotation {
```

Comment:
I think you could make `ANSIAnnotation` a top-level the and nest `ANSIColor` and `ANSITrait` in it. It somehow feels weird to me to nest this in the `DiagnosticsFormatter`.

Reply:
Agree with this as well. What made me go for nesting was that I wasn't sure how useful this "supporting" type would be to other parts of the codebase but I'm more than happy to decouple it. Slight bike-shedding question but does that warrant its own file or should it stay within `DiagnosticFormatter.swift`? If I'm nesting the color and trait types anyways, would you agree that the `ANSI` prefix becomes kinda superfluous and that it can be dropped?

Reply:
I don't have strong opinions since I don't anticipate any further users of it in SwiftDiagnostics. I'll leave it up to you.